File: scripts/fetch_all_lotto.js
/**
 * 동행복권 전회차(1회 ~ 1240회) 고속 수집기 (Node.js 비동기)
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const OUTPUT_PATH = path.join(__dirname, '..', 'src', 'data', 'lotto_history.json');
const BASE_URL = 'https://www.dhlottery.co.kr/lt645/selectPstLt645Info.do?srchLtEpsd=';

function parseDate(str) {
  const s = String(str);
  if (s.length === 8) {
    return `${s.slice(0, 4)}-${s.slice(4, 6)}-${s.slice(6, 8)}`;
  }
  return s;
}

async function fetchDraw(drwNo) {
  for (let retry = 0; retry < 3; retry++) {
    try {
      const res = await fetch(`${BASE_URL}${drwNo}`, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
          'Referer': 'https://www.dhlottery.co.kr/'
        },
        signal: AbortSignal.timeout(6000)
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const json = await res.json();
      const list = json?.data?.list;
      if (list && list.length > 0) {
        const item = list[0];
        return {
          drwNo: item.ltEpsd,
          date: parseDate(item.ltRflYmd),
          numbers: [
            item.tm1WnNo,
            item.tm2WnNo,
            item.tm3WnNo,
            item.tm4WnNo,
            item.tm5WnNo,
            item.tm6WnNo
          ],
          bonus: item.bnsWnNo,
          firstWinamnt: item.rnk1WnAmt || 0,
          firstPrzwnerCo: item.rnk1WnNope || 0,
          totSellamnt: item.wholEpsdSumNtslAmt || 0
        };
      }
      return null;
    } catch (e) {
      await new Promise(r => setTimeout(r, 400));
    }
  }
  return null;
}

async function main() {
  console.log('최신 회차 확인 중...');
  let latestDrw = 1240;
  // 1241 확인
  const nextCheck = await fetchDraw(1241);
  if (nextCheck) latestDrw = 1241;
  console.log(`확인된 최신 회차: ${latestDrw}회`);

  const dir = path.dirname(OUTPUT_PATH);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  const existingMap = new Map();
  if (fs.existsSync(OUTPUT_PATH)) {
    try {
      const prev = JSON.parse(fs.readFileSync(OUTPUT_PATH, 'utf-8'));
      for (const d of prev.draws || []) {
        existingMap.set(d.drwNo, d);
      }
      console.log(`기존 파일에서 ${existingMap.size}건 로드됨.`);
    } catch (e) {}
  }

  const targets = [];
  for (let i = 1; i <= latestDrw; i++) {
    if (!existingMap.has(i)) targets.push(i);
  }

  console.log(`신규 수집 대상: ${targets.length}건`);

  // 동시 10개씩 배치 처리
  const BATCH_SIZE = 10;
  let finished = 0;

  for (let i = 0; i < targets.length; i += BATCH_SIZE) {
    const chunk = targets.slice(i, i + BATCH_SIZE);
    const results = await Promise.all(chunk.map(no => fetchDraw(no)));
    for (const item of results) {
      if (item) existingMap.set(item.drwNo, item);
    }
    finished += chunk.length;
    if (finished % 100 === 0 || finished === targets.length) {
      console.log(`진행률: ${finished}/${targets.length} (${((finished / targets.length) * 100).toFixed(1)}%)`);
    }
    await new Promise(r => setTimeout(r, 100)); // 서버 배려
  }

  const sortedDraws = Array.from(existingMap.values()).sort((a, b) => a.drwNo - b.drwNo);
  const payload = {
    lastUpdated: new Date().toISOString().replace('T', ' ').slice(0, 19),
    latestDrawNo: latestDrw,
    totalCount: sortedDraws.length,
    draws: sortedDraws
  };

  fs.writeFileSync(OUTPUT_PATH, JSON.stringify(payload, null, 2), 'utf-8');
  console.log(`✅ 완료! 총 ${sortedDraws.length}건 저장 완료: ${OUTPUT_PATH}`);
}

main().catch(console.error);

File: scripts/fetch_all_lotto.py
"""
동행복권 신규 API(selectPstLt645Info.do)를 활용한 역대 전회차(1회~최신회차) 수집 스크립트
"""
import os
import json
import time
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed

BASE_URL = "https://www.dhlottery.co.kr/lt645/selectPstLt645Info.do?srchLtEpsd={}"
OUTPUT_PATH = os.path.join(os.path.dirname(os.path.dirname(__file__)), "src", "data", "lotto_history.json")

def parse_date(date_str):
    # '20241214' -> '2024-12-14'
    if len(date_str) == 8:
        return f"{date_str[:4]}-{date_str[4:6]}-{date_str[6:]}"
    return date_str

def fetch_draw(drw_no):
    url = BASE_URL.format(drw_no)
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
        "Referer": "https://www.dhlottery.co.kr/"
    }
    req = urllib.request.Request(url, headers=headers)
    for attempt in range(3):
        try:
            with urllib.request.urlopen(req, timeout=5) as resp:
                data = json.loads(resp.read().decode('utf-8'))
                items = data.get("data", {}).get("list", [])
                if items and len(items) > 0:
                    row = items[0]
                    return {
                        "drwNo": row["ltEpsd"],
                        "date": parse_date(str(row["ltRflYmd"])),
                        "numbers": [
                            row["tm1WnNo"],
                            row["tm2WnNo"],
                            row["tm3WnNo"],
                            row["tm4WnNo"],
                            row["tm5WnNo"],
                            row["tm6WnNo"]
                        ],
                        "bonus": row["bnsWnNo"],
                        "firstWinamnt": row.get("rnk1WnAmt", 0),
                        "firstPrzwnerCo": row.get("rnk1WnNope", 0),
                        "totSellamnt": row.get("wholEpsdSumNtslAmt", 0)
                    }
                return None
        except Exception:
            time.sleep(0.3)
    return None

def main():
    os.makedirs(os.path.dirname(OUTPUT_PATH), exist_ok=True)
    
    # 1. 최신 회차 찾기
    print("최신 회차 확인 중...")
    curr = 1240
    while True:
        res = fetch_draw(curr + 1)
        if res:
            curr += 1
        else:
            break
    
    latest_drw = curr
    print(f"확인된 최신 회차: {latest_drw}회")

    existing_data = {}
    if os.path.exists(OUTPUT_PATH):
        try:
            with open(OUTPUT_PATH, "r", encoding="utf-8") as f:
                saved = json.load(f)
                for item in saved.get("draws", []):
                    existing_data[item["drwNo"]] = item
            print(f"기존 저장된 데이터: {len(existing_data)}건")
        except Exception:
            pass

    targets = [i for i in range(1, latest_drw + 1) if i not in existing_data]
    print(f"신규 수집 대상 회차: {len(targets)}건")

    if targets:
        count = 0
        with ThreadPoolExecutor(max_workers=15) as executor:
            future_to_drw = {executor.submit(fetch_draw, drw): drw for drw in targets}
            for future in as_completed(future_to_drw):
                res = future.result()
                if res:
                    existing_data[res["drwNo"]] = res
                count += 1
                if count % 100 == 0 or count == len(targets):
                    print(f"수집 진행: {count}/{len(targets)} ({count/len(targets)*100:.1f}%)")

    all_draws = [existing_data[k] for k in sorted(existing_data.keys())]

    output_payload = {
        "lastUpdated": time.strftime("%Y-%m-%d %H:%M:%S"),
        "latestDrawNo": latest_drw,
        "totalCount": len(all_draws),
        "draws": all_draws
    }

    with open(OUTPUT_PATH, "w", encoding="utf-8") as f:
        json.dump(output_payload, f, ensure_ascii=False, indent=2)

    print(f"완료: 1회부터 {latest_drw}회까지 총 {len(all_draws)}건이 {OUTPUT_PATH}에 성공적으로 저장되었습니다.")

if __name__ == "__main__":
    main()
